### DISCUSSION TITLE
5v to 3.3v signal conversion

### DISCUSSION THREAD

**Discussion by DeeEmm** [DIY-Flow-Bench/DIY-Flow-Bench#91] [Shield Design]
With the change in processor to the ESP32 also comes a change in the electronics side of things.

If you have put together a shield or breadboarded a design for the old Arduino based system, you can still use it but you will need to do some small modifications as the ESP is a 3.3v device whereas the Uno/Mega that we previously used was a 5V device.

The big change is that the the MAF and pressure sensors are all generally 5v devices, that is they require a 5v supply and output a 0-5v signal. Some MAF sensors actually require a 12v supply.

A voltage divider on the 5V output will do the job, but they can be inaccurate as there will be tolerances on the advertised values. You can to a degree overcome this by actually measuring the resistors you use and selecting the most appropriate values to give you the best result. You can also use trimpots instead of fixed values to dial the value in. You can find out the actual output voltage by placing your divider across 5v and measuring the actual output. 

There is also the option to use the software trimpot in the configuration file to add an offset to the value. (voltage dividers are linear in nature)

Regular logic level shifters will not work for analog signals so you have to use something like an opamp. This is not as simple as  using a voltage divider but is more accurate.

I'm also looking at using some ADC's that use I2C. These are 12bit and so are a really good match for the ESP32. Using I2C also free's up the I/O which is currently at its limit. ADCs are relatively simple to integrate. THis will require some code changes to implement.

Based on this you can also buy a pre-packaged MPXV7007DP pressure sensor mounted on a breadboard with an MCP3221 ADC. These are made by MIKROE and known as a [Diff-Press-Click](https://www.mikroe.com/diff-press-click). Prices do however vary greatly but with the inflated prices of MPXV7007DP sensors it's probably not too bad. 

I might consider adding support for these to the shield rather than having pressure sensors direct mounted. This way if you don't need the additional pressure sensors, you don't have to use them. They simply plug in to a regular pin header socket.

For a MAF based system you don't actually need the reference pressure to be able to measure mass flow, but you do need it to convert to volumetric flow and also to convert between reference pressures.

**Comment by DeeEmm** [comment]
Some more info on the MIKROE MIKROE-3332
- https://libstock.mikroe.com/projects/view/2665/diff-press-click
- https://download.mikroe.com/documents/standards/mikrobus/mikrobus-standard-specification-v200.pdf
- https://download.mikroe.com/documents/add-on-boards/click/diff-press/diff-press-schematic-v100.pdf
- https://www.mikroe.com/diff-press-click

**Comment by KirikauKiwi** [comment]
I received my ads1115 today, it has a way of achieving the 3.3V logic level for the I2c whilst having a 5V supply voltage. It could be worth looking at, as it may possibly eliminate the need for the logic shifter on the current v2.3 shield for later revisions.

The beak out board is from [DFROBOT](https://www.dfrobot.com/product-1730.html) and here is the [schematic](https://www.dfrobot.com.cn/images/upload/File/20180529102659fma5x2.pdf)

**Reply by DeeEmm** [reply]
All SDL / SDA lines have to be held high for I2C comms to work. 

My understanding of the ADS1115 device is that SDL and SDA will be at VCC, which in our case is 5v. Using pull-ups to 3.3v on the SDA/SCL lines will not limit the I2C signals to 3.3v, they will still rise to 5v when the ADS1115 is communicating on the bus. This is the reason that the level converter was used. 

I suspect that 3.3v has been used by DFROBOT so that their BOB will work on both 3.3v and 5v systems, as 3.3v is above the threshold for logical '1'

**Reply by KirikauKiwi** [reply]
Looking at the schematic if ads1115 is powered by 5V to the VDD and the 3.3V LDO is tied to the I2C lines with an 10kOhm pulldown resistor on each leg, it would be about a 1.7V voltage drop, 5V - 1.7V = 3.3V. Essentially the 5V I2C line will drop to 3.3V

**Reply by DeeEmm** [reply]
When the ADS1115 pulls the bus high, the voltage will be pulled up to 5v. (vcc). The 10k pull-up to the 3.3v line will pull it back down a little, but it will not keep it at 3.3v, it will be somwhere in between 3.3 and 5v. The actual value depends on the internal resistance of the ADS1115. We can take a wild stab that it is a large value of R similar to the pull-up...

<img width="1016" alt="Screenshot 2024-11-03 at 4 26 45 pm" src="https://github.com/user-attachments/assets/68e4201b-e521-4540-b362-04f06d6c9cd5">

This gives us 4.15v, which exceed the 3.3v limit of the ESP32. Equally the internal resistance of the ADS1115 could be much lower than 10k, in which case the bus would be even higher than the 4.15v shown.

At any rate. 3.3v is the limit for the ESP32, so anything higher than this is potentially damaging. 

The issue with relying on things like voltage dividers to do logic bus shifting is that it is fraught with risk. It could work in *some* specific cases, but we cannot guarantee that the end user will select the correct components at the right tolerances or use the specific BOB that is known to work. My experience is that people tend to use anything but the recommended components. lol.

The other consideration is that we are not dealing with component level PCB design here. So we have no control over what components people will actually use. This increases the risk that the end user will have an issue, which in turn increases support requirements and reduces end user satisfaction. 

IMHO, the proper way to connect non-matching I2C busses when using modular components as we are here, is to use a logic level shifter like the PCA9306. If the end user want to omit it in favour of using a voltage divider, then that is of course their choice.

**Reply by KirikauKiwi** [reply]
Thanks for the through conversation, I would tend to agree with you here in using a using a logic level shifter being the better way. I will check what the actual voltage is on the DFROBOT board I have as it says it is a 3.3V logic level in its feature and don't really want it potentially frying a pin.
